### PR TITLE
Fix for axisymmetric terms in NEMO + general NEMO updates

### PR DESCRIPTION
**Pull request by WallyMaier**
## Proposed Changes
-This work is focused on fixing some of the axisymmetric loop structures within NEMO, which prevented residual evolution.   This also changes a viscous planar wedge regression to an axisymmetric case.  

-Code updates to make NEMO more-in-line with SU2 code style.  i.e. variable declarations, SU2 MPI errors, etc.

-Fixes bug in vibrational-electronic functionality for CSU2TCLib

-Updating ionization/nheavy loops so CSU2TCLib and CMutationTCLib are consistent.   

## Related Work
This addresses/builds on #1106 #1205 

## PR Checklist
- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [x] My contribution is commented and consistent with SU2 style.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (WallyMaier) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:1668]: This is more consistent with the baseline SU2.

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (pcarruscag) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:836]: That is unlikely, the NEMO solver does not declare itself to be OpenMP-ready (via method GetHasHybridParallel() -> true) and so it will not be called multi-threaded by the Integration classes.
Also if you do not compile with -Dwith-omp=enabled (or true, I can never recall) all those SU2_OMP_? macros are no-ops.

Comment (WallyMaier) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:836]: So this was a super complex problem to solve....
I had the gradient computation inside the for loop.  So as you can imagine, things went south quite quickly.

Comment (pcarruscag) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:836]: Actually you did point me to an hybrid parallel bug...

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:841]:
```suggestion
    SU2_OMP_FOR_STAT(omp_chunk_size)
    for (iPoint = 0; iPoint < nPoint; iPoint++) {
```

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:853]:
```suggestion
   }
   END_SU2_OMP_FOR
```

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp]: Move this to CFVMFlowSolverBase.inl and share it between CEulerSolver and CNEMOEulerSolver please (the bug is in CEuler solver, not here)

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp
Comment (pcarruscag) [inline SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:367]: Have a look at #1312, I think there was a logic error here with how "Vector" is used which I tried to fix.

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (pcarruscag) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp]: Yes, the logic is more or less: if the method does not have to be a template then it is declared in the .hpp and implemented in .inl of CFVMSolverBase.
If it has to be a template, or you want it to be "inlineable", then it has to go in the .hpp.
But both files are headers, the only thing that changes is the "visibility", e.g. CEulerSolver includes the hpp but not the inl.
At the moment the inl is included by the NS solvers, at the top of the associated .cpp's you'll see
```c++
/*--- Explicit instantiation of the parent class of CIncEulerSolver,
 *    to spread the compilation over two cpp files. ---*/
template class CFVMFlowSolverBase<CIncEulerVariable, ENUM_REGIME::INCOMPRESSIBLE>;
```
That also means that if you change something in the .inl only the NS solvers recompile, but not the Euler solvers (which sounded like a good idea because e.g. CEulerSolver is a reasonably large file).

The incompressible version looks different enough to be either its own method or left as is.
My rule of thumb for such master versions is not to have physics-based conditionals in there (e.g. "if compressible"), if necessary, that type of detail should be provided from the outside, and not already present in the body of the "master" version.
In other words, we want those versions to be generic in the sense they allow externally provided specialization, not in the sense they already try to cover all possible options.
I hope this makes some sense.

File: SU2_CFD/src/numerics/flow/flow_sources.cpp
Comment (FlorianDm) [reply SU2_CFD/src/numerics/flow/flow_sources.cpp:159]: Indeed I believe it should actually be +k dT/dy then. Thanks for spotting this. The test case i made breaks so I will rerun it with this branch and add the change to the PR

Comment (FlorianDm) [reply SU2_CFD/src/numerics/flow/flow_sources.cpp:159]: Ok so I reran the test case with this branch. I guess I should wait until this PR is to be merged to push the changes in the regression files to the TestCases repo so any other open PRs don't get failed regressions

Comment (FlorianDm) [reply SU2_CFD/src/numerics/flow/flow_sources.cpp:159]: or is the rest ready to go? If it may take a while maybe I better do the fix now in another PR in concert with the TestCases changes so it can be merged straight away as a simple fix and then you just merge the latest develop in here.. let me know what you think is best

Comment (WallyMaier) [reply SU2_CFD/src/numerics/flow/flow_sources.cpp:159]: @FlorianDm Glad you could confirm!  This PR should be done on Monday/Tuesday (hopefully) and Ill update the regression tests.  However, if you want to go ahead to address it case before then, thats fine as well!

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:770]: Note to myself: This should be a negative, I negative should change in ComputeMixtureEnergies.

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (WallyMaier) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp]: Done!

**Comment by WallyMaier** [comment]
Everything is up-to-date. I just need to update the TestCase folder....not sure what that order is...

File: TestCases/hybrid_regression.py
Comment (FlorianDm) [inline TestCases/hybrid_regression.py:230]: Just to clarify, the way I had done the regressions was with RESTART=YES in the config file where the restart file in TestCases solution_flow.dat was already converged to rho<1e-12 so it would just do 10 iterations and then those values I took. I see you did it differently since the values are only at 1e-6. I guess it does not matter but just wanted to clarify

File: SU2_CFD/include/solvers/CFVMFlowSolverBase.hpp
Comment (pcarruscag) [inline SU2_CFD/include/solvers/CFVMFlowSolverBase.hpp:283]:
```suggestion
  void ComputeAxisymmetricAuxGradients(CGeometry *geometry, const CConfig* config);
```
I'm not aware of the proper naming but let's go for something more descriptive.

File: SU2_CFD/include/solvers/CFVMFlowSolverBase.inl
Comment (pcarruscag) [inline SU2_CFD/include/solvers/CFVMFlowSolverBase.inl:2928]:
```suggestion
void CFVMFlowSolverBase<V, FlowRegime>::ComputeAxisymmetricAuxGradients(CGeometry *geometry, const CConfig* config) {
```
Nicely done! Thanks.

File: SU2_CFD/src/solvers/CEulerSolver.cpp
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CEulerSolver.cpp:2342]:
```suggestion
      ComputeAxisymmetricAuxGradients(geometry, config);
```

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:840]:
```suggestion
    ComputeAxisymmetricAuxGradients(geometry,config);
```

**Review by pcarruscag — APPROVED** [review]

**Review by pcarruscag — COMMENTED** [review]
The order is to merge into testcases first just before merging this branch (then you can manually trigger the tests to re-run).

File: TestCases/hybrid_regression.py
Comment (WallyMaier) [reply TestCases/hybrid_regression.py:230]: @FlorianDm, thanks for the clarification.  I only updated the values in the regression test, so it still should be a restart.

**Pull request merged**
